src/middleware.js

```diff
@@ -56,27 +57,49 @@ export function withCarBlockHandler (handler) {
 }
 
 /**
- * Creates a dagula instance backed by the R2 blockstore backed by content claims.
+ * Creates a dagula instance backed by content claims.
  *
  * @type {import('@web3-storage/gateway-lib').Middleware<BlockContext & DagContext & UnixfsContext & IndexSourcesContext & IpfsUrlContext, IndexSourcesContext & IpfsUrlContext, Environment>}
  */
 export function withContentClaimsDagula (handler) {
   return async (request, env, ctx) => {
     const { dataCid } = ctx
-    const index = new ContentClaimsIndex(asSimpleBucket(env.CARPARK), {
+    const locator = ContentClaimsLocator.create({
       serviceURL: env.CONTENT_CLAIMS_SERVICE_URL ? new URL(env.CONTENT_CLAIMS_SERVICE_URL) : undefined
     })
-    const found = await index.get(dataCid)
-    if (!found) {
-      // fallback to old index sources and dagula fallback
-      return composeMiddleware(
-        withIndexSources,
-        withDagulaFallback
-      )(handler)(request, env, ctx)
+    const locRes = await locator.locate(dataCid.multihash)
+    if (locRes.error) {
+      if (locRes.error.name === 'NotFound') {
+        // fallback to old index sources and dagula fallback
+        return composeMiddleware(
+          withIndexSources,
+          withDagulaFallback
+        )(handler)(request, env, ctx)
+      }
+      throw new Error(`failed to locate: ${dataCid}`, { cause: locRes.error })
     }
-    const blockstore = new BatchingR2Blockstore(env.CARPARK, index)
 
-    const dagula = new Dagula(blockstore)
+    const fetcher = BatchingFetcher.create(locator)
+    const dagula = new Dagula({
+      async get (cid) {
+        const res = await fetcher.fetch(cid.multihash)
+        return res.ok ? { cid, bytes: await res.ok.bytes() } : undefined
+      },
+      async stream (cid, options) {
+        const res = await fetcher.fetch(cid.multihash, options)
+        // @ts-expect-error `MINIFLARE` is not a property of `globalThis`
+        if (globalThis.MINIFLARE && res.ok) {
+          return res.ok.stream().pipeThrough(new TransformStream({
+            transform: (chunk, controller) => controller.enqueue(chunk.slice())
```

Comment:
can we document code path reasoning here?

Reply:
ha, yes...it's this: https://github.com/web3-storage/freeway/blob/1c7b7a6faff866bb7c8fece427777e33e7e27416/src/lib/blockstore.js#L193-L220